modAionImpl/src/org/aion/zero/impl/db/PendingBlockStore.java

```diff
@@ -0,0 +1,816 @@
+     * if persistence is requested but not achievable.
+     *
+     * @param _props properties of the databases to be used for storage
+     * @throws InvalidFilePathException when given a persistent database vendor for which the path
+     *     cannot be created
+     */
+    public PendingBlockStore(final Properties _props) throws InvalidFilePathException {
+        Properties local = new Properties(_props);
+
+        // check for database persistence requirements
+        DBVendor vendor = DBVendor.fromString(local.getProperty(Props.DB_TYPE));
+        if (vendor.getPersistence()) {
+            File pbFolder =
+                    new File(local.getProperty(Props.DB_PATH), local.getProperty(Props.DB_NAME));
+
+            verifyAndBuildPath(pbFolder);
+            local.setProperty(Props.DB_PATH, pbFolder.getAbsolutePath());
+        }
+
+        init(local);
+    }
+
+    /**
+     * Initializes and opens the databases where the pending blocks will be stored.
+     *
+     * @param props the database properties to be used in initializing the underlying databases
+     */
+    private void init(Properties props) {
+        // initialize status
+        this.status = new HashMap<>();
+
+        // create the level source
+        props.setProperty(Props.DB_NAME, LEVEL_DB_NAME);
+        this.levelDatabase = connectAndOpen(props, LOG);
+        this.levelSource = new ObjectDataSource<>(levelDatabase, HASH_LIST_RLP_SERIALIZER);
+
+        // create the queue source
+        props.setProperty(Props.DB_NAME, QUEUE_DB_NAME);
+        this.queueDatabase = connectAndOpen(props, LOG);
+        this.queueSource = new ObjectDataSource<>(queueDatabase, BLOCK_LIST_RLP_SERIALIZER);
+
+        // create the index source
+        props.setProperty(Props.DB_NAME, INDEX_DB_NAME);
+        this.indexSource = connectAndOpen(props, LOG);
+    }
+
+    /**
+     * Checks that the underlying storage was correctly initialized and open.
+     *
+     * @return true if correctly initialized and the databases are open, false otherwise.
+     */
+    public boolean isOpen() {
+        return status != null
+                && levelSource.isOpen()
+                && queueSource.isOpen()
+                && indexSource.isOpen();
+    }
+
+    private static final Serializer<List<byte[]>, byte[]> HASH_LIST_RLP_SERIALIZER =
+            new Serializer<>() {
+                @Override
+                public byte[] serialize(List<byte[]> object) {
+                    byte[][] infoList = new byte[object.size()][];
+                    int i = 0;
+                    for (byte[] b : object) {
+                        infoList[i] = RLP.encodeElement(b);
+                        i++;
+                    }
+                    return RLP.encodeList(infoList);
+                }
+
+                @Override
+                public List<byte[]> deserialize(byte[] stream) {
+                    RLPList list = (RLPList) RLP.decode2(stream).get(0);
```

Comment:
That is how `decode2` works. It returns a list where the first element contains the decoding.

Reply:
Incorrect serializations are unlikely to occur since the class controls both the serializer and deserializer and even null objects will be serialized correctly.